Base node liveliness consistently on gossiper::is_alive

bhalevy opened the pull request:
Currently he gossiper marks endpoint_state objects as alive/dead.
I some cases the endpoint_state::is_alive function is checked but in many other cases
gossiper::is_alive(endpoint) is used to determine if the endpoint is alive.

This series removed the endpoint_state::is_alive state and moves all the logic to gossiper::is_alive
that bases its decision on the endpoint having an endpoint_state and being in the _live_endpoints set.

For that, the _live_endpoints is made sure to be replicated to all shards when changed
and the endpoint_state changes are serialized under lock_endpoint, and also making sure that the
endpoint_state in the _endpoint_states_map is never updated in place, but rather a temporary copy is changed
and then safely replicated using gossiper::replicate

Refs https://github.com/scylladb/scylladb/issues/14794

bhalevy commented:
Converted to draft until all relevant dtests pass (currently, there are some disagreements with:
```
FAILED update_cluster_layout_tests.py::TestUpdateClusterLayout::test_remove_garbage_members_from_group0_after_abort_decommission[leaving_token_ring-]
FAILED update_cluster_layout_tests.py::TestUpdateClusterLayout::test_remove_garbage_members_from_group0_after_abort_decommission[Announcing_that_I_have_left_the_ring-]
```

And some individual changes were submitted as independent PRs, so waiting for those to be merged (#14797, #14798)

asias commented:
> Currently he gossiper marks endpoint_state objects as alive/dead. I some cases the endpoint_state::is_alive function is checked but in many other cases gossiper::is_alive(endpoint) is used to determine if the endpoint is alive.
> 
> This series removed the endpoint_state::is_alive state and moves all the logic to gossiper::is_alive that bases its decision on the endpoint having an endpoint_state and being in the _live_endpoints set.
> 
> For that, the _live_endpoints is made sure to be replicated to all shards when changed and the endpoint_state changes are serialized under lock_endpoint, and also making sure that the endpoint_state in the _endpoint_states_map is never updated in place, but rather a temporary copy is changed and then safely replicated using gossiper::replicate

Does this series fix any real bug?  I not sure why it is better to use _live_endpoints instead of the is_alive property of the endpoint_state object?

It's hard to review a ~50 commits series. Better to split things like marking const to a follow up series can focus on the minimal change to reduce the noise.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2633/

bhalevy commented:
> > Currently he gossiper marks endpoint_state objects as alive/dead. I some cases the endpoint_state::is_alive function is checked but in many other cases gossiper::is_alive(endpoint) is used to determine if the endpoint is alive.
> > This series removed the endpoint_state::is_alive state and moves all the logic to gossiper::is_alive that bases its decision on the endpoint having an endpoint_state and being in the _live_endpoints set.
> > For that, the _live_endpoints is made sure to be replicated to all shards when changed and the endpoint_state changes are serialized under lock_endpoint, and also making sure that the endpoint_state in the _endpoint_states_map is never updated in place, but rather a temporary copy is changed and then safely replicated using gossiper::replicate
> 
> Does this series fix any real bug?

It doesn't fix any known customer/QA issue.
But I do consider the fact that the endpoint_state may be inconsistent across shards and the is_alive state inconsistent between `eps->is_alive()` and `gossiper::is_alive(ep)` an issue that needs fixing.

> I not sure why it is better to use _live_endpoints instead of the is_alive property of the endpoint_state object?

The is_alive property requires modifying an endpoint_state object in place, probably on shard 0 and it's unclear how and when it propagates to other shards that replicate this endpoint state.

By basing the state on presence in `_live_endpoints` this series ensures that the state is applied consistently on all shards.
Some additional locking was added (and the lack of them can be considered other issue(s) that are fixed here) to also ensure that the state changes are serialized properly so gossiper doesn't end up with a mix of states from racing state updates.

> 
> It's hard to review a ~50 commits series. Better to split things like marking const to a follow up series can focus on the minimal change to reduce the noise.

BTW, some of the const changes are related to making get_endpoint_state_for_endpoint_ptr and get_endpoint_state return const ptr/ref, while introducing `get_mutable_endpoint_state_ptr` for private gossiper use - to prevent in-place modification of the endpoint_state while it's in the map - to facilitate exception safety via `replicate()`

tgrabiec commented:
Please change the title to something more meaningful (a verb). "Gossiper live endpoints" doesn't say what is accomplished, especially that we already have gossiper live endpoints.

bhalevy commented:
In v2 ([a440dde](https://github.com/scylladb/scylladb/commit/a440dde8a23a5e3291da8a7c194c3455bcfc20f4)):
- rebased and cleaned up series

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2923/

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -1660,16 +1661,16 @@ future<> gossiper::handle_major_state_change(inet_address ep, const endpoint_sta
     if (!is_dead_state(ep_state)) {
         mark_alive(ep, ep_state);
+
+        if (eps_old) {
```
How is this code movement related to the rest of the commit?

bhalevy replied:
It can actually be refactored out.
It's wrong to notify `on_restart` if the new state `is_dead_state`.

kbr-scylla replied:
Why?

And what if the node went through multiple state transitions: it first restarted in non-dead state, then went to a dead state? Is it possible that we missed the non-dead state moment, and only observed the later moment in dead state, but it is still necessary to call `on_restart` due to that intermediate state that we failed to observe? (But other nodes could have)

kbr-scylla replied:
Generally I'd like to avoid changes like this when there is no known bug that is being fixed, while they might have some unexpected consequences later.

bhalevy replied:
fair enough.  will drop and make a note to investigate this as a follow up .
the whole on_restart handling in storage_service:;on_restart seems fishy...

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -2275,14 +2289,19 @@ bool gossiper::is_alive(inet_address ep) const {
     if (ep == get_broadcast_address()) {
         return true;
     }
+    bool is_alive = _live_endpoints.contains(ep);
```
`gossiper::is_alive` is on the hot paths, won't this regress performance?

kbr-scylla replied:
Still, it's two `unordered_set` searches instead of one. So assuming that the `unordered_set` search takes a significant portion of the operation (I don't know if that's the case, but we should consider it), we would increase that significant portion twice.

I'm quite worried about this change.

But wait. Do we actually need the `get_endpoint_state_for_endpoint_ptr` call below? Maybe we can just replace one search with another.

We should maintain an invariant: `_live_endpoints` is always a subset of `_endpoint_state_map` (which is used by `get_endpoint_state_for_endpoint_ptr`) (maybe the invariant is already satisfied? If not -- why?)

And then the second part of this function would be redundant, so we could have only a single search.

bhalevy replied:
The invariant should be satisfied, although I didn't audit the code to make sure it is.
FWIW, I've never seen the warning below...

We can do the check as a safety belt only in non-release builds so it won't hurt performance in production but would still be checked in testing.

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -2275,14 +2289,19 @@ bool gossiper::is_alive(inet_address ep) const {
     // badness.  (See CASSANDRA-1463 for an example).
     if (eps) {
-        return eps->is_alive();
+        auto is_alive_state = this->is_alive_state(*eps);
```
We should probably call `is_alive_state` only when DEBUG logging is enabled

bhalevy replied:
Will do

bhalevy replied:
Actually, a later patch deletes the hunk, so I'll just get rid of it in this patch.

bhalevy commented on service/migration_manager.cc:
```diff
@@ -222,7 +222,7 @@ void migration_manager::schedule_schema_pull(const gms::inet_address& endpoint,
 
 bool migration_manager::have_schema_agreement() {
-    const auto known_endpoints = _gossiper.get_endpoint_states();
+    const auto known_endpoints = _gossiper.get_live_endpoint_states();
```
True.  It's not a problem, but I can document that in the change log.
In a follow up series to change the gossiper interface to support host_id also `get_endpoint_states` is going to return a copy of the endpoint states, although in the future model the endpoint_state is going to be kept in a lw_shared_ptr so the copy of the map can still point to the shared state, reducing the cost.

asias replied:
This is a unnecessary change for this big enough series. If you want to avoid eps.is_alive() just call g.is_alive(). 

I do not think it worth adding such expensive function.

bhalevy replied:
I think that a `get_live_endpoint_states` can be more efficient than calling g.is_alive for every endpoint since it can simply walk `_live_endpoints` and get (a lw_shared_ptr of) the endpoint state for each endpoint, but w ecan indeed leave it to a follow-up for brevity sake.

asias replied:
The have_schema_agreement is used only in two places. Both of them are in slow path. I'd suggest drop get_live_endpoint_states for now.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -680,7 +680,7 @@ future<> gossiper::remove_endpoint(inet_address endpoint, permit_id pid) {
         logger.info("removed {} from _seeds, updated _seeds list = {}", endpoint, _seeds);
```
oops, not sure how it ended up like that.
It's `O(log |nodes|)`

kbr-scylla replied:
It should actually become amortized `O(1)` since it's an unordered set, not `std::set`

bhalevy commented on gms/gossiper.cc:
```diff
@@ -1380,10 +1380,14 @@ endpoint_state& gossiper::get_endpoint_state(inet_address ep) {
 }
 
 future<> gossiper::reset_endpoint_state_map() {
```
It's a public interface so I it's not very elegant to assume it will be called only on shard 0.
But `invoke_on` on the same shard is inexpensive.  It doesn't actually perform a cross-shard task.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -681,8 +681,8 @@ future<> gossiper::remove_endpoint(inet_address endpoint, permit_id pid) {
     _live_endpoints.erase(endpoint);
-    co_await update_live_endpoints_version();
     _unreachable_endpoints.erase(endpoint);
+    co_await replicate_live_endpoints_on_change();
```
The motivation is that for relying on `_live_endpoints` for `gossiper::is_alive` we want `_live_endpoints` to be updated on all shards.  Today that happens in some undetermined future time when `replicate_live_endpoints_on_change` is called.  This change makes sure that any change to `_live_endpoints` is propagated to all other shards as it's made. I'll amend the change log respectively.

> Also the function has a precondition which the code doesn't satisfy now:

Good point. I'll address that.

bhalevy commented:
In v3 ([03c8308](https://github.com/scylladb/scylladb/commit/03c8308ccc28b3364fcca42439200835b5f9dd7c)):
- rebased + retested
- gossiper: add get_live_endpoint_states
  - improved commit message
- gossiper: add lock_endpoint_update_semaphore
- gossiper: reset_endpoint_state_map: clear live/unreachable endpoints on all shards
  - assume shard 0
  - lock_endpoint_update_semaphore
- gossiper: fold update_live_endpoints_version into replicate_live_endpoints_on_change
  - add calls to lock_endpoint_update_semaphore
- gossiper: handle_major_state_change: notify on_restart only if !is_dead_state
  - refactored out into a separate patch
- gossiper: is_alive: use _live_endpoints
  - deleted debug code

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2955/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2954/

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -1614,12 +1611,14 @@ future<> gossiper::real_mark_alive(inet_address addr) {
     } else {
         _endpoints_to_talk_with.front().push_back(addr);
     }
+    lock.return_all();
```
How about using a scope instead?
```cpp
{
    auto lock = co_await lock_endpoint_update_semaphore();
    co_await replicate_live_endpoints_on_change();
    if (_endpoints_to_talk_with.empty()) {
        _endpoints_to_talk_with.push_back({addr});
    } else {
        _endpoints_to_talk_with.front().push_back(addr);
    }
}
```

(I see you use a scope in the other place)

bhalevy replied:
Next version introduces a mutate function that acquires the lock, bumps the version if needed and calls replicate...on_change

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -582,9 +582,9 @@ future<> gossiper::do_apply_state_locally(gms::inet_address node, const endpoint
-                    logger.trace("Ignoring remote version {} <= {} for {}", remote_max_version, local_max_version, node);
+                    logger.debug("Ignoring remote version {} <= {} for {}", remote_max_version, local_max_version, node);
                 }
-                if (!local_state.is_alive() && !this->is_dead_state(local_state)) { // unless of course, it was dead
```
We don't need to check `this->is_alive(node)` in place of the previous check?

bhalevy replied:
It shouldn't be a problem to call mark_alive on an endpoint that's already alive
but saving that makes sense.

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -711,7 +711,7 @@ future<> gossiper::do_status_check() {
         }
         auto& ep_state = *eps;
-        bool is_alive = ep_state.is_alive();
+        bool is_dead = is_dead_state(ep_state);
```
I don't get this change.

`is_dead_state` is not the negation of `is_alive`, a node could be dead but not in DEAD_STATES. So I don't think it's correct to use `is_dead` in place of where `!is_alive` was used earlier (like below with the `expire_time` check)

Shouldn't we use `!is_alive(endpoint)`?

bhalevy replied:
Fair point.  I'll change to `is_alive = _live_endpoints.contains(endpoint)` which is the equivalent after this series of what `ep_state.is_alive()` was.

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -1138,7 +1134,7 @@ int64_t gossiper::get_endpoint_downtime(inet_address ep) const noexcept {
 // It is called from failure_detector
 future<> gossiper::convict(inet_address endpoint) {
     auto* state = get_endpoint_state_for_endpoint_ptr(endpoint);
-    if (!state || !state->is_alive()) {
```
Why get rid of the check? (shouldn't we use `is_alive(endpoint)` now instead?)

bhalevy replied:
Later version of this series brought the check back.
It is indeed needed to prevent convict from repeatedly calling `mark_dead` for the endpoint.

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -1633,7 +1627,7 @@ future<> gossiper::real_mark_alive(inet_address addr) {
 future<> gossiper::mark_dead(inet_address addr, endpoint_state& local_state, permit_id pid) {
     logger.trace("marking as down {}", addr);
-    local_state.mark_dead();
+    auto permit = co_await lock_endpoint(addr, pid);
```
A sneaked-in bugfix? Should probably be mentioned in commit message.

bhalevy replied:
hmm, good catch.
It somehow escaped https://github.com/scylladb/scylladb/pull/14845
I submit it as a separate patch, or even a separate PR.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -895,8 +895,9 @@ future<> gossiper::failure_detector_loop() {
-                std::set_difference(nodes.begin(), nodes.end(), _live_endpoints.begin(), _live_endpoints.end(), std::back_inserter(nodes_down));
+                auto live_nodes = boost::copy_range<std::vector<inet_address>>(_live_endpoints);
+                std::sort(live_nodes.begin(), live_nodes.end());
+                std::set_difference(nodes.begin(), nodes.end(), live_nodes.begin(), live_nodes.end(), std::back_inserter(nodes_down));
```
let's consider that as a followup, since I don't want to mess with the node ordering here in this series.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -946,6 +941,7 @@ future<> gossiper::replicate_live_endpoints_on_change() {
 
         if (live_endpoint_changed) {
             _shadow_live_endpoints = _live_endpoints;
+            ++_live_endpoints_version;
```
> Just an idea, not sure if worth pursuing.

I'm trying to keep the scope of this series small.
There are indeed many ideas for improvement.
Let's consider that as a followup...

My main goal is to proceed to the next step which to make the changes to the endpoint state values atomic  and exception safe by preparing the modified endpoint_state maps in temporary variables instead of in place as it is done in many places today, and then let `replicate()` apply them in an exception safe way on all shards or on none.

Then the next step is extend the endpoint states model to support dual-indexing, by inet_address and by host_id and providing a set of entry point to get to the endpoint state by host_id, and converge all users to use them, and eventually drop the address-based entry points we won't use.

bhalevy commented on service/storage_service.cc:
```diff
@@ -3267,7 +3267,7 @@ future<> storage_service::on_dead(gms::inet_address endpoint, gms::endpoint_stat
     slogger.debug("endpoint={} on_restart: permit_id={}", endpoint, pid);
     // If we have restarted before the node was even marked down, we need to reset the connection pool
-    if (state.is_alive()) {
+    if (endpoint != get_broadcast_address() && _gossiper.is_alive(endpoint)) {
```
This is needed, IIRC when replacing a node with the same ip address.
`gossiper::is_alive` always returns true for `get_broadcast_address()`
but previously I think the endpoint_state was not marked with is_alive at this point yet.

I'd be more than happy to drop on_restart as a whole BTW, if possible...

bhalevy commented on gms/gossiper.cc:
```diff
@@ -1388,13 +1388,13 @@ endpoint_state& gossiper::get_endpoint_state(inet_address ep) {
 future<> gossiper::reset_endpoint_state_map() {
-    if (this_shard_id() != 0) {
-        on_internal_error(logger, "must be called on shard 0");
-    }
```
Removed the "transient" check in the next version

bhalevy commented on gms/gossiper.cc:
```diff
@@ -1388,13 +1388,13 @@ endpoint_state& gossiper::get_endpoint_state(inet_address ep) {
+    logger.debug("Resetting endpoint state map");
+    auto lock = lock_endpoint_update_semaphore();
+    auto version = _live_endpoints_version + 1;
+    co_await container().invoke_on_all([version] (gossiper& g) {
```
> Clearing them now would get rid of potential redundant replication on the next gossiper round.

More than that. It can actually prevent replication by mistake if _live_endpoints is updated to become equal to _shadow_live_endpoints (as was after the shadow round), so it's an actual bug.

Opened https://github.com/scylladb/scylladb/issues/15003

bhalevy commented:
In v4 ([7192525](https://github.com/scylladb/scylladb/commit/7192525c7cbbb241860fc4e8d551566c080823f1)):
- rebased  + retested
- gossiper: make _live_endpoints an unordered_set
  - Fixed change log regarding O(1) complexity
- gossiper: functions that change _live_endpoints must be called on shard 0
  - dropped on_internal_error (that was deleted in a later patch)
- gossiper: reset_endpoint_state_map: clear also shadow endpoint sets
  - Fixes #15003
- gossiper: add mutate_live_and_unreachable_endpoints
  - to safely update live/unreachable endpoints, including bumping of live_endpoints_version and replication on change
- gossiper: fold update_live_endpoints_version into replicate_live_endpoints_on_change
  - Use mutate_live_and_unreachable_endpoints
  - Do not bump version in `replicate_live_endpoints_on_change`
- gossiper: evict_from_membership: erase endpoint from _live_endpoints
  - erase also from _live_endpoints and issue on_internal_error if the endpoint was in _live_endpoints unexpectedly
- dropped "gossiper: handle_major_state_change: notify on_restart only if !is_dead_state"
- gossiper: is_alive: use _live_endpoints
  - check only presence in _live_endpoints in production
  - sanity test presence in endpoint states map in non-release modes - and issue on_internal_error if node is alive but has no endpoint_state
- endpoint_state: get rid of _is_alive member and methods
  - keep check for `!is_alive(node)` in `do_apply_state_locally` before calling `mark_alive`

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -675,15 +675,17 @@ future<> gossiper::remove_endpoint(inet_address endpoint, permit_id pid) {
-    _live_endpoints.erase(endpoint);
-    co_await update_live_endpoints_version();
-    _unreachable_endpoints.erase(endpoint);
+    co_await mutate_live_and_unreachable_endpoints([endpoint] (gossiper& g) {
```
Aren't we now calling `lock_endpoint_update_semaphore()` inside `mutate_...` while the lock is already being held (we took it above)? Doesn't this deadlock?

bhalevy replied:
oops, sorry, the lock is removed in the next patch by mistake.  Likely due to a pilot error while editing the series.  I'll rebase and make sure it's gone at this stage.

bhalevy replied:
fixed in v5

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -820,6 +820,13 @@ future<semaphore_units<>> gossiper::lock_endpoint_update_semaphore() {
+future<> gossiper::mutate_live_and_unreachable_endpoints(std::function<void(gossiper&)> func) {
+    auto lock = co_await lock_endpoint_update_semaphore();
+    func(*this);
+    _live_endpoints_version += (_live_endpoints != _shadow_live_endpoints);
```
We could increment it unconditionally too, it wouldn't hurt.

bhalevy replied:
Initially I did bump it unconditionally, but since it would cause an unnecessary "hiccup" of the failure detector if only _unreachable_endpoints in modified (in `evict_from_membership` and possibly in `real_mark_alive`), I decided to make it more accurate (BTW bumping it in `replicate_live_endpoints_on_change` in the previous version did just that - bump it only on change from shadow_live_endpoints).

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -1174,6 +1173,12 @@ future<> gossiper::evict_from_membership(inet_address endpoint, permit_id pid) {
     co_await mutate_live_and_unreachable_endpoints([endpoint] (gossiper& g) {
         g._unreachable_endpoints.erase(endpoint);
+        // When evict_from_membership is called, the endpoint should have
+        // already been erased from _live_endpoints, so enforce that here.
```
Are we 100% sure that must indeed be the case?

I'm imagining a decommissioning node, let's say it successfully decommissioned and shutdown, and now we evict it from membership before we managed to mark it as dead. I don't think that would be a problem, the two events can happen "at the same time".

bhalevy replied:
`evict_from_membership` is called either explicitly after `remove_endpoint`, which erases the endpoint from live_endpoints, or after the endpoint has expired in _expire_time_endpoint_map.
It inserted there by `advertise_token_removed` -> `add_expire_time_for_endpoint` which happens when the node has left the ring, so it's not supposed to be alive in this state.

That said, we don't have to strict in this case and just erase it unconditionally from _live_endpoints and not care if it was there or not.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2992/

bhalevy commented:
In v5 ([8ce2d80](https://github.com/scylladb/scylladb/commit/8ce2d806b71eca390634cb7a51b1a1e5b008fd6d)):
- rebased
- gossiper: fold update_live_endpoints_version into replicate_live_endpoints_on_change
  - drop the call to `lock_endpoint_update_semaphore` before `mutate_live_and_unreachable_endpoints` (that was dropped in a later patch in v4, so no functional change overall)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3091/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3091/

https://github.com/scylladb/scylla-dtest/pull/3509

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3093/

avikivity commented:
@asias please review

bhalevy commented:
In v6 ([6f7efc1](https://github.com/scylladb/scylladb/commit/6f7efc1f10151e0cdf21b718e0cd3bfa80ecbadc)):
- rebase
- gossiper: add _unreachable_endpoints_version
- gossiper: add mutate_live_and_unreachable_endpoints
  - move responsibility to bump _{live,unreachable}_endpoints_version to callers
- gossiper: fold update_live_endpoints_version into replicate_live_endpoints_on_change
  - bump _{live,unreachable}_endpoints_version
- gossiper:: replicate_live_endpoints_on_change: use versions to detect change
- gossiper:: replicate_live_endpoints_on_change: fixup indentation
- endpoint_state: get rid of _is_alive member and methods
  - get rid of capture of _endpoint_state_map in replicate_live_endpoints_on_change
  - keep `is_alive` check in `gossiper::convict` (it is meaningful)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3144/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3144/

I'm not sure why the MV test become flaky in this branch, but as far as I could tell, the issue is with the test that reads the system_distributed.view_build_status table with cl=ONE and so it misses updates done on some nodes early in the startup process withcl=ONE, and so they don't get to other nodes (and since hints are disabled they require repair).

The fix is to use higher consistency level (ALL or QUORUM, when possible).
I submitted https://github.com/scylladb/scylla-dtest/pull/3517 with this fix and an improvement to the test to verify `wait_for_view` robustness.

bhalevy commented:
@asias please review

asias commented on gms/gossiper.cc:
```diff
@@ -1393,10 +1393,13 @@ endpoint_state& gossiper::get_endpoint_state(inet_address ep) {
-    co_await container().invoke_on_all([] (gossiper& g) {
+    logger.debug("Resetting endpoint state map");
+    auto lock = lock_endpoint_update_semaphore();
+    auto version = _live_endpoints_version + 1;
```
The version is updated before _live_endpoints is really updated. I think it is safer to always update the _live_endpoints before update the _live_endpoints_version.

bhalevy replied:
But here `_live_endpoints_version` is not changed, we just calculate the next value (under the lock).
The new value is set below, after clearing the live and unreachable endpoints sets.

asias replied:
You are right. Sorry, I misread it.

asias commented on gms/gossiper.hh:
```diff
@@ -244,8 +245,20 @@ private:
+    future<semaphore_units<>> lock_endpoint_update_semaphore();
+
+    // Must be called on shard 0.
+    // Safely update _live_endpoints and/or _unreachable_endpoints
```
I want to mention that even with mutate_live_and_unreachable_endpoints, we still could have inconsistency of _live_endpoints on difference shards. Yes, after this patch, we modify _live_endpoints under the lock. But we do not read _live_endpoints under the lock. So different shards could have different view of  _live_endpoints.

bhalevy replied:
True, whoever needs momentary cross-shard consistency can acquire the lock (like `get_live_members_synchronized`)

bhalevy replied:
okay

asias commented on gms/gossiper.cc:
```diff
@@ -1047,7 +1047,7 @@ void gossiper::run() {
 
             {
-                auto lock = get_units(_endpoint_update_semaphore, 1).get();
+                auto lock = lock_endpoint_update_semaphore().get();
```
Do we still need to call replicate_live_endpoints_on_change in the gossip loop here? I think we do not need it anymore because every time _live_endpoints is accessed,  we replicate them to all shards already.

bhalevy replied:
You're right. This isn't needed after using `mutate_live_and_unreachable_endpoints`.
I'll add a patch to remove it.

bhalevy commented:
In v7 ([cc23f0e](https://github.com/scylladb/scylladb/commit/cc23f0e7dcb9cfaffd9b07411f78058b78ca6af6)):
- rebased
- dropped "gossiper: add get_live_endpoint_states"
  - use __gossiper.is_alive instead

asias commented on gms/gossiper.hh:
```diff
@@ -228,6 +228,7 @@ private:
 
     /* unreachable member set */
     std::unordered_map<inet_address, clk::time_point> _unreachable_endpoints;
+    uint64_t _unreachable_endpoints_version = 0;
```
I think we do not really need to introduce _unreachable_endpoints_version. In almost every cases, we always update _live_endpoints_version and the new  _unreachable_endpoints_version together. 

Node being live and unreachable is a rare operation. It is better to make the logic simple than optimize over calling a few unreachable nodes.

bhalevy replied:
The simplest way would be to always replicate both in today's `replicate_live_endpoints_on_change` (that could be renamed to `replicate_live_and_unreachable_endpoints` and bump `_live_endpoints_version` unconditionally there.

asias replied:
Yes, bumping the version but _live or _unreachable endpoints does not change is not a big deal.

asias commented on gms/gossiper.cc:
```diff
@@ -931,42 +942,30 @@ future<> gossiper::replicate_live_endpoints_on_change() {
+            if (local_gossiper._shadow_unreachable_endpoints_version != _unreachable_endpoints_version) {
+                local_gossiper._unreachable_endpoints = _unreachable_endpoints;
+                local_gossiper._unreachable_endpoints_version = _unreachable_endpoints_version;
+                local_gossiper._shadow_unreachable_endpoints_version = _unreachable_endpoints_version;
```
I think the above replicating logic can be simplified and we can remove the whole shadow thing.
```
if (local_gossiper._live_endpoints_version = _live_endpoints_version) {
      local_gossiper._live_endpoints  =  _live_endpoints
      local_gossiper._unreachable_endpoints =  _unreachable_endpoints 
      local_gossiper._live_endpoints_version =  _live_endpoints_version
}
```

We can check if the non zero shard needs updating by checking if the non zero shard has the same version.
The replicating runs under the lock, so there is no danger that  the _live_endpoints or _unreahable_endpoints would change while shard zero is replicating to different non zero shards. 

Btw, like I mentioned in the other thread, we can combine live and shadow version.

bhalevy replied:
Since in this branch replicate_live_endpoints_on_change is called only under mutate_live_and_unreachable_endpoints it can simply bump the _live_endpoints_version and replicate _live_endpoints, _unreachable_endpoints, and _live_endpoints_version unconditionally.

asias replied:
Yes! After the removal of the replicate_live_endpoints_on_change from gossiper::run.

bhalevy replied:
I was over-optimistic. I forgot it is also called from get_live_members_synchronized, and in this case we don't want to replicate unconditionally, so next version will just use _live_endpoints_version to detect a change, and it is bumped unconditionally in mutate_live_and_unreachable_endpoints to force an update on all other shards.

asias commented on gms/gossiper.cc:
```diff
@@ -1166,7 +1166,17 @@ version_type gossiper::get_max_endpoint_state_version(endpoint_state state) cons
+        // already been erased from _live_endpoints, so enforce that here.
+        auto was_live = g._live_endpoints.erase(endpoint);
+        if (was_live) {
+            on_internal_error_noexcept(logger, fmt::format("Node {} was unexpectedly marked as live when evicted from membership", endpoint));
```
This is a more generic issue than node to be evicted should be not in live_endpoints. It is the issue that we need to make sure the _live_endpoints are in sync with _endpoint_state_map. 

Perhaps we can add some check in gossiper::run() to make sure _live_endpoints and _endpoint_state_map are in sync. Since now we reply much more on the _live_endpoints than before. 

I'd suggest drop the on_internal_error_noexcept because it is not a serious problem that deserve an abort, we are removing the node anyway.

bhalevy replied:
okay.  the consistency of endpoint_state with _live_endpoints is also checked (in this branch) in gossiper::is_alive in non-release modes.
It is not symmetric as we want to require the all node in _live_endpoints have an endpoint_state, but not vice verse.
It is possible to have an endpoint_state of a node that has left the ring and is not in _live_endpoints.

bhalevy commented:
In v8 ([801987a](https://github.com/scylladb/scylladb/commit/801987ab19a5b0f2eedefd59c02a58301257b7db)):
- gossiper: add mutate_live_and_unreachable_endpoints
  - Unconditionally bump _live_endpoints_version
  - Add comment in header file regarding exception safety
- gossiper: fold update_live_endpoints_version into replicate_live_endpoints_on_change
  - Simplified.
  - _live_endpoints_version is now managed centrally in mutate_live_and_unreachable_endpoints
- gossiper: run: no need to replicate live_endpoints
- gossiper: replicate_live_endpoints_on_change: use _live_endpoints_version to detect change
  - simplified
  - Use only _live_endpoints_version to detect a change per-shard
  - Updated last to protect against exceptions on shard.
- gossiper: evict_from_membership: erase endpoint from _live_endpoints
  - simplified
  - erase from _live_endpoints unconditionally
  - with no warning nor abort if evicted endpoint was in _live_endpoints
-

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3184/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3208/

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -947,44 +947,32 @@ future<> gossiper::replicate_live_endpoints_on_change() {
     //
     // Gossiper task runs only on CPU0:
     //
-    //    - If endpoint_state_map or _live_endpoints have changed - replicate
```
The "shadow" copies were used to safely replicate these variables to other shards, while allowing the original variables to be modified on shard 0 in parallel (avoiding a data race).

Do I understand correctly that right now the parallel modification possibility is gone, because we always modify and replicate under the lock?

bhalevy replied:
correct

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -947,44 +947,32 @@ future<> gossiper::replicate_live_endpoints_on_change() {
-                }
-
-                for (auto&& e : es) {
-                    local_gossiper._endpoint_state_map[e.first].set_alive(e.second.is_alive());
```
Re commit order: In theory this hunk should only be applied after we change `is_alive` to use `_live_endpoints`, IIUC.

kbr-scylla replied:
Ah sorry, I see it's above, missed it. nevermind
